openwisp_controller/config/static/config/js/widget.js

```diff
@@ -153,6 +153,50 @@
         return error;
     };
 
+    var getAllVariables = function () {
+        var systemDefinedVariables = JSON.parse($('#system_context').html()),
+            configVariables;
+        if (window.getContext()) {
+            configVariables = JSON.parse(window.getContext().value);
+        }
+        var allVariables = Object.assign({}, systemDefinedVariables, configVariables);
+        return allVariables;
+    };
+
+    const debounce = (delay, func) => {
+        var debounceTimer;
+        return function() {
+            clearTimeout(debounceTimer);
+            debounceTimer = setTimeout(() => func.apply(this, arguments), delay);
+        };
+    };
+
+    var validateMaxLength = function() {
+        var allVariables = getAllVariables(),
+            variablesKeys = Object.keys(allVariables),
+            allFields = $(".jsoneditor .vTextField");
+        allFields.map((index, field) => {
+            if ($(field).is(":visible") && $(field).attr('maxLength')) {
+                var maxLength = $(field).attr('maxLength');
+                $(field).off('input');
+                $(field).on('input', debounce(2000, () => {
```

Comment:
2000 is 2 seconds and is probably too much.
As long as this code is triggered once for every keystroke we'll be fine.